src/riak_kv_qry_compiler.erl

```diff
@@ -170,27 +182,33 @@ compile_select_clause(DDL, ?SQL_SELECT{'SELECT' = #riak_sel_clause_v1{ clause =
     %% whole query
     Acc = {sets:new(), #riak_sel_clause_v1{ }},
     %% iterate from the right so we can append to the head of lists
-    {ResultTypeSet, Q2} = lists:foldl(CompileColFn, Acc, Sel),
+    {ResultTypeSet, Sel1} = lists:foldl(CompileColFn, Acc, Sel),
 
     {ColTypes, Errors} = my_mapfoldl(
         fun(ColASTX, Errors) ->
             infer_col_type(DDL, ColASTX, Errors)
         end, [], Sel),
 
-    case sets:is_element(aggregate, ResultTypeSet) of
-        true  ->
-            Q3 = Q2#riak_sel_clause_v1{
+    IsGroupBy = (Q?SQL_SELECT.group_by /= []),
+    IsAggregate = sets:is_element(aggregate, ResultTypeSet),
+    if
+        IsGroupBy ->
+            Sel2 = Sel1#riak_sel_clause_v1{
+                   calc_type = group_by,
+                   col_names = get_col_names(DDL, Q) };
```

Comment:
This should also set the initial state for `group by` which is currently being done in `riak_kv_qry_worker:execute_query/2` at line 210